// @gmod/vcf/dist/parse.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var vcfReserved_1 = __importDefault(require("./vcfReserved"));
function Variant(stuff) {
    //@ts-ignore
    Object.assign(this, stuff);
}
function decodeURIComponentNoThrow(uri) {
    try {
        return decodeURIComponent(uri);
    }
    catch (e) {
        // avoid throwing exception on a failure to decode URI component
        return uri;
    }
}
/**
 * Class representing a VCF parser, instantiated with the VCF header.
 * @param {object} args
 * @param {string} args.header - The VCF header. Supports both LF and CRLF
 * newlines.
 * @param {boolean} args.strict - Whether to parse in strict mode or not (default true)
 */
var VCF = /** @class */ (function () {
    function VCF(_a) {
        var _this = this;
        var _b = _a.header, header = _b === void 0 ? '' : _b, _c = _a.strict, strict = _c === void 0 ? true : _c;
        if (!header || !header.length) {
            throw new Error('empty header received');
        }
        var headerLines = header.split(/[\r\n]+/).filter(function (line) { return line; });
        if (!headerLines.length) {
            throw new Error('no non-empty header lines specified');
        }
        this.strict = strict;
        this.metadata = JSON.parse(JSON.stringify({
            INFO: vcfReserved_1.default.InfoFields,
            FORMAT: vcfReserved_1.default.GenotypeFields,
            ALT: vcfReserved_1.default.AltTypes,
            FILTER: vcfReserved_1.default.FilterTypes,
        }));
        var lastLine;
        headerLines.forEach(function (line) {
            if (!line.startsWith('#')) {
                throw new Error("Bad line in header:\n".concat(line));
            }
            else if (line.startsWith('##')) {
                _this._parseMetadata(line);
            }
            else {
                lastLine = line;
            }
        });
        if (!lastLine) {
            throw new Error('No format line found in header');
        }
        var fields = lastLine.trim().split('\t');
        var thisHeader = fields.slice(0, 8);
        var correctHeader = [
            '#CHROM',
            'POS',
            'ID',
            'REF',
            'ALT',
            'QUAL',
            'FILTER',
            'INFO',
        ];
        if (fields.length < 8) {
            throw new Error("VCF header missing columns:\n".concat(lastLine));
        }
        else if (fields.length === 9) {
            throw new Error("VCF header has FORMAT but no samples:\n".concat(lastLine));
        }
        else if (thisHeader.length !== correctHeader.length ||
            !thisHeader.every(function (value, index) { return value === correctHeader[index]; })) {
            throw new Error("VCF column headers not correct:\n".concat(lastLine));
        }
        this.samples = fields.slice(9);
    }
    VCF.prototype._parseGenotypes = function (format, prerest) {
        var _this = this;
        var rest = prerest.split('\t');
        var genotypes = {};
        var formatKeys = format === null || format === void 0 ? void 0 : format.split(':');
        if (formatKeys) {
            this.samples.forEach(function (sample, index) {
                genotypes[sample] = {};
                formatKeys.forEach(function (key) {
                    genotypes[sample][key] = null;
                });
                rest[index]
                    .split(':')
                    .filter(function (f) { return f; })
                    .forEach(function (val, index) {
                    var thisValue;
                    if (val === '' || val === '.' || val === undefined) {
                        thisValue = null;
                    }
                    else {
                        var entries = val
                            .split(',')
                            .map(function (ent) { return (ent === '.' ? null : ent); });
                        var valueType = _this.getMetadata('FORMAT', formatKeys[index], 'Type');
                        if (valueType === 'Integer' || valueType === 'Float') {
                            thisValue = entries.map(function (val) { return (val ? +val : val); });
                        }
                        else {
                            thisValue = entries;
                        }
                    }
                    genotypes[sample][formatKeys[index]] = thisValue;
                }, {});
            });
        }
        return genotypes;
    };
    /**
     * Parse a VCF metadata line (i.e. a line that starts with "##") and add its
     * properties to the object.
     * @param {string} line - A line from the VCF. Supports both LF and CRLF
     * newlines.
     */
    VCF.prototype._parseMetadata = function (line) {
        var match = line.trim().match(/^##(.+?)=(.*)/);
        if (!match) {
            throw new Error("Line is not a valid metadata line: ".concat(line));
        }
        var _a = match.slice(1, 3), metaKey = _a[0], metaVal = _a[1];
        if (metaVal.startsWith('<')) {
            if (!(metaKey in this.metadata)) {
                this.metadata[metaKey] = {};
            }
            var _b = this._parseStructuredMetaVal(metaVal), id = _b[0], keyVals = _b[1];
            this.metadata[metaKey][id] = keyVals;
        }
        else {
            this.metadata[metaKey] = metaVal;
        }
    };
    /**
     * Parse a VCF header structured meta string (i.e. a meta value that starts
     * with "<ID=...")
     * @param {string} metaVal - The VCF metadata value
     *
     * @returns {Array} - Array with two entries, 1) a string of the metadata ID
     * and 2) an object with the other key-value pairs in the metadata
     */
    VCF.prototype._parseStructuredMetaVal = function (metaVal) {
        var keyVals = this._parseKeyValue(metaVal.replace(/^<|>$/g, ''), ',');
        var id = keyVals.ID;
        delete keyVals.ID;
        if ('Number' in keyVals) {
            if (!Number.isNaN(Number(keyVals.Number))) {
                keyVals.Number = Number(keyVals.Number);
            }
        }
        return [id, keyVals];
    };
    /**
     * Get metadata filtered by the elements in args. For example, can pass
     * ('INFO', 'DP') to only get info on an metadata tag that was like
     * "##INFO=<ID=DP,...>"
     * @param  {...string} args - List of metadata filter strings.
     *
     * @returns {any} An object, string, or number, depending on the filtering
     */
    VCF.prototype.getMetadata = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var filteredMetadata = this.metadata;
        for (var i = 0; i < args.length; i += 1) {
            filteredMetadata = filteredMetadata[args[i]];
            if (!filteredMetadata) {
                return filteredMetadata;
            }
        }
        return filteredMetadata;
    };
    /**
     * Sometimes VCFs have key-value strings that allow the separator within
     * the value if it's in quotes, like:
     * 'ID=DB,Number=0,Type=Flag,Description="dbSNP membership, build 129"'
     *
     * Parse this at a low level since we can't just split at "," (or whatever
     * separator). Above line would be parsed to:
     * {ID: 'DB', Number: '0', Type: 'Flag', Description: 'dbSNP membership, build 129'}
     * @param {string} str - Key-value pairs in a string
     * @param {string} [pairSeparator] - A string that separates sets of key-value
     * pairs
     *
     * @returns {object} An object containing the key-value pairs
     */
    VCF.prototype._parseKeyValue = function (str, pairSeparator) {
        if (pairSeparator === void 0) { pairSeparator = ';'; }
        var data = {};
        var currKey = '';
        var currValue = '';
        var state = 1; // states: 1: read key to = or pair sep, 2: read value to sep or quote, 3: read value to quote
        for (var i = 0; i < str.length; i += 1) {
            if (state === 1) {
                // read key to = or pair sep
                if (str[i] === '=') {
                    state = 2;
                }
                else if (str[i] !== pairSeparator) {
                    currKey += str[i];
                }
                else if (currValue === '') {
                    data[currKey] = null;
                    currKey = '';
                }
            }
            else if (state === 2) {
                // read value to pair sep or quote
                if (str[i] === pairSeparator) {
                    data[currKey] = currValue;
                    currKey = '';
                    currValue = '';
                    state = 1;
                }
                else if (str[i] === '"') {
                    state = 3;
                }
                else {
                    currValue += str[i];
                }
            }
            else if (state === 3) {
                // read value to quote
                if (str[i] !== '"') {
                    currValue += str[i];
                }
                else {
                    state = 2;
                }
            }
        }
        if (state === 2 || state === 3) {
            data[currKey] = currValue;
        }
        else if (state === 1) {
            data[currKey] = null;
        }
        return data;
    };
    /**
     * Parse a VCF line into an object like { CHROM POS ID REF ALT QUAL FILTER
     * INFO } with SAMPLES optionally included if present in the VCF
     * @param {string} line - A string of a line from a VCF. Supports both LF and
     * CRLF newlines.
     */
    VCF.prototype.parseLine = function (line) {
        var _this = this;
        // eslint-disable-next-line no-param-reassign
        line = line.trim();
        if (!line.length) {
            return undefined;
        }
        //@ts-ignore
        var parser = this; // so we can include this in lazy-property closure
        var currChar = 0;
        for (var currField = 0; currChar < line.length; currChar += 1) {
            if (line[currChar] === '\t') {
                currField += 1;
            }
            if (currField === 9) {
                // reached genotypes, rest of fields are evaluated lazily
                break;
            }
        }
        var fields = line.substr(0, currChar).split('\t');
        var rest = line.substr(currChar + 1);
        var CHROM = fields[0], POS = fields[1], ID = fields[2], REF = fields[3], ALT = fields[4], QUAL = fields[5], FILTER = fields[6];
        var chrom = CHROM;
        var pos = +POS;
        var id = ID === '.' ? null : ID.split(';');
        var ref = REF;
        var alt = ALT === '.' ? null : ALT.split(',');
        var qual = QUAL === '.' ? null : +QUAL;
        var filter = FILTER === '.' ? null : FILTER.split(';');
        if (this.strict && fields[7] === undefined) {
            throw new Error("no INFO field specified, must contain at least a '.' (turn off strict mode to allow)");
        }
        var info = fields[7] === undefined || fields[7] === '.'
            ? {}
            : this._parseKeyValue(fields[7]);
        Object.keys(info).forEach(function (key) {
            var items;
            if (info[key]) {
                items = info[key]
                    .split(',')
                    .map(function (val) { return (val === '.' ? null : val); })
                    .map(function (f) { return (f ? decodeURIComponentNoThrow(f) : f); });
            }
            else {
                // it will be falsy so just assign whatever is there
                items = info[key];
            }
            var itemType = _this.getMetadata('INFO', key, 'Type');
            if (itemType) {
                if (itemType === 'Integer' || itemType === 'Float') {
                    items = items.map(function (val) {
                        if (val === null) {
                            return null;
                        }
                        return Number(val);
                    });
                }
                else if (itemType === 'Flag') {
                    if (info[key]) {
                        // eslint-disable-next-line no-console
                        console.warn("Info field ".concat(key, " is a Flag and should not have a value (got value ").concat(info[key], ")"));
                    }
                    else {
                        items = true;
                    }
                }
            }
            info[key] = items;
        });
        //@ts-ignore
        var variant = new Variant({
            CHROM: chrom,
            POS: pos,
            ALT: alt,
            INFO: info,
            REF: ref,
            FILTER: filter && filter.length === 1 && filter[0] === 'PASS' ? 'PASS' : filter,
            ID: id,
            QUAL: qual,
        });
        Object.defineProperty(variant, 'SAMPLES', {
            get: function () {
                var samples = parser._parseGenotypes(fields[8], rest);
                Object.defineProperty(this, 'SAMPLES', {
                    value: samples,
                    configurable: false,
                });
                return samples;
            },
            configurable: true,
        });
        //@ts-ignore
        return variant;
    };
    return VCF;
}());
exports.default = VCF;
